python/ray/dag/tests/experimental/test_mocked_nccl_dag.py

```diff
@@ -0,0 +1,355 @@
+    with InputNode() as inp:
+        dag = sender.send_dict.bind(inp)
+        dag = dag.with_type_hint(
+            TorchTensorType(transport="nccl", _static_tensor_schema=True)
+        )
+        dag = receiver.recv_dict.bind(dag)
+
+    compiled_dag = dag.experimental_compile()
+
+    shape = 10
+    dtype = torch.float16
+    # Test sending a dictionary that always contains keys inserted in the same
+    # order, but with different-sized and different-valued tensors.
+    # same keys.
+    for i in range(3):
+        tensor_shape = (shape * (i + 1),)
+        spec = [
+            ("a", i, tensor_shape, dtype),
+            ("b", i, tensor_shape, dtype),
+            ("c", i, tensor_shape, dtype),
+        ]
+        ref = compiled_dag.execute(spec)
+        assert ray.get(ref) == spec
+
+    ray.kill(barrier1)
+    ray.kill(barrier2)
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster",
+    [
+        {
+            "num_cpus": 2,
+            "num_gpus": 2,
+            "num_nodes": 1,
+        }
+    ],
+    indirect=True,
+)
+def test_p2p_static_tensor_schema_error(ray_start_cluster):
+    """
+    Test that when static_tensor_schema=True, an error is thrown when a
+    different number of tensors is found.
+    """
+    # Barrier name should be barrier-{sender rank}-{receiver rank}.
+    # Create a barrier in both directions because we don't know which rank will
+    # get assigned to sender and receiver.
+    barrier1 = Barrier.options(name="barrier-0-1").remote()  # noqa
+    barrier2 = Barrier.options(name="barrier-1-0").remote()  # noqa
+
+    sender = MockedWorker.remote()
+    receiver = MockedWorker.remote()
+
+    ray.get([sender.start_mock.remote(), receiver.start_mock.remote()])
+
+    # Test torch.Tensor sent between actors.
+    with InputNode() as inp:
+        dag = sender.send_dict.bind(inp)
+        dag = dag.with_type_hint(
+            TorchTensorType(transport="nccl", _static_tensor_schema=True)
+        )
+        dag = receiver.recv_dict.bind(dag)
+
+    compiled_dag = dag.experimental_compile()
+
+    shape = 10
+    dtype = torch.float16
+    # Test sending a dictionary that always contains keys inserted in the same
+    # order, but with different-sized and different-valued tensors.
+    # same keys.
+    for i in range(3):
+        tensor_shape = (shape * (i + 1),)
+        spec = [
+            ("a", i, tensor_shape, dtype),
+            ("b", i, tensor_shape, dtype),
+            ("c", i, tensor_shape, dtype),
+        ]
+        ref = compiled_dag.execute(spec)
+        assert ray.get(ref) == spec
+
+    # Test sending a dictionary with a different number of entries.
+    spec = [
+        ("a", 1, shape, dtype),
+        ("b", 1, shape, dtype),
+    ]
+    ref = compiled_dag.execute(spec)
+    with pytest.raises(OSError, match="Channel closed"):
+        ray.get(ref)
+
+    # Sending correct shape still errors because the DAG has already been torn
```

Comment:
Yeah it's a limitation of this technique :( not sure if there is a way for NCCL to recover from the bad recv.

Reply:
Actually this could probably be improved by storing the exception returned by the actor task as the DAG result and then restarting the DAG. But too complicated for now, will leave a TODO.

Reply:
can you add docstring to this argument that any failure can cause dag to be teardown>?